tests/test_facebook_bookmarks.py

```diff
@@ -43,73 +45,224 @@ def expected_sync_streams():
+            calculated_state_formatted = datetime.datetime.strftime(calculated_state_as_datetime, state_format)
+
+            stream_to_calculated_state[stream] = calculated_state_formatted
+
+        return stream_to_calculated_state
 
-    def get_properties(self):  # pylint: disable=arguments-differ
-        return {'start_date' : '2015-03-15T00:00:00Z',
-                'account_id': os.getenv('TAP_FACEBOOK_ACCOUNT_ID'),
-                'end_date': '2015-03-16T00:00:00+00:00',
-                'insights_buffer_days': '1'
-        }
 
     def test_run(self):
-        conn_id = connections.ensure_connection(self)
+        all_streams =  self.expected_sync_streams()
+        non_insight_streams = {stream for stream in all_streams if not self.is_insight(stream)}
+        insight_streams = {stream for stream in all_streams if self.is_insight(stream)}
+
+        # Testing against ads insights objects
+        self.start_date = self.get_properties()['start_date']
+        self.end_date = self.get_properties()['end_date']
+        self.bookmarks_test(insight_streams)
+
+        # Testing agianst core objects
+        self.end_date = '2021-02-09T00:00:00Z'
+        self.bookmarks_test(non_insight_streams)
+
+
+    def bookmarks_test(self, expected_streams):
+        """A Parametrized Bookmakrs Test"""
+        expected_replication_keys = self.expected_replication_keys()
+        expected_replication_methods = self.expected_replication_method()
+        expected_insights_buffer = -1 * int(self.get_properties()['insights_buffer_days'])  # lookback window
+
+        ##########################################################################
+        ### First Sync
+        ##########################################################################
+
+        conn_id = connections.ensure_connection(self, original_properties=False)
+
+        # Run in check mode
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # Select only the expected streams tables
+        catalog_entries = [ce for ce in found_catalogs if ce['tap_stream_id'] in expected_streams]
+        self.perform_and_verify_table_and_field_selection(conn_id, catalog_entries, select_all_fields=True)
+
+        # Run a sync job using orchestrator
+        first_sync_record_count = self.run_and_verify_sync(conn_id)
+        first_sync_records = runner.get_records_from_target_output()
+        first_sync_bookmarks = menagerie.get_state(conn_id)
+
+        ##########################################################################
+        ### Update State Between Syncs
+        ##########################################################################
+
+        new_states = {'bookmarks': dict()}
+        simulated_states = self.calculated_states_by_stream(first_sync_bookmarks)
+        for stream, new_state in simulated_states.items():
+            replication_key = list(expected_replication_keys[stream])[0]
```

Comment:
Couldn't this be avoided if `calculated_states_by_stream` hands back the new state? It already knows what the replication key for each stream is, right?

Reply:
Yeah, I can do that. It makes more sense your way.

Reply:
Done.